core/src/replay_stage.rs

```diff
@@ -149,6 +149,62 @@ struct SkippedSlotsInfo {
     last_skipped_slot: u64,
 }
 
+struct PartitionInfo {
+    partition_exists: bool,
+    partition_start_time: Option<Instant>,
+}
+
+impl PartitionInfo {
+    fn new() -> Self {
+        Self {
+            partition_exists: false,
+            partition_start_time: None,
+        }
+    }
+
+    fn update(
+        &mut self,
+        partition_detected: bool,
+        heaviest_bankslot: Slot,
+        last_voted_slot: Slot,
+        reset_bank_slot: Slot,
+    ) {
+        if !self.partition_exists && partition_detected {
+            warn!("PARTITION DETECTED waiting to join heaviest fork: {} last vote: {:?}, reset slot: {}",
+                heaviest_bankslot,
+                last_voted_slot,
+                reset_bank_slot,
+            );
+            datapoint_info!(
+                "replay_stage-partition-start",
+                ("heaviest_slot", heaviest_bankslot as i64, i64),
+                ("last_vote_slot", last_voted_slot as i64, i64),
+                ("reset_slot", reset_bank_slot as i64, i64)
+            );
+            self.partition_exists = true;
+            self.partition_start_time = Some(Instant::now());
+        } else if self.partition_exists && !partition_detected {
+            warn!(
+                "PARTITION resolved heaviest fork: {} last vote: {:?}, reset slot: {}",
+                heaviest_bankslot, last_voted_slot, reset_bank_slot
+            );
+            datapoint_info!(
+                "replay_stage-partition-resolved",
+                ("heaviest_slot", heaviest_bankslot as i64, i64),
+                ("last_vote_slot", last_voted_slot as i64, i64),
+                ("reset_slot", reset_bank_slot as i64, i64),
+                (
+                    "partition_duration_secs",
+                    self.partition_start_time.unwrap().elapsed().as_secs() as i64,
```

Comment:
do we want millisecond granularity for small to to three slot partitions?

Reply:
Extra granularity won't kill us. I actually started with milliseconds and switched to seconds. I'll switch back